apps/taxes/src/modules/taxjar/taxjar-order-created-adapter.ts

```diff
@@ -0,0 +1,105 @@
+import { CreateOrderRes, LineItem } from "taxjar/dist/util/types";
+import { OrderCreatedSubscriptionFragment } from "../../../generated/graphql";
+import { ChannelConfig } from "../channels-configuration/channels-config";
+import { numbers } from "../taxes/numbers";
+import { CreateOrderResponse } from "../taxes/tax-provider-webhook";
+import { WebhookAdapter } from "../taxes/tax-webhook-adapter";
+import { CreateOrderArgs, TaxJarClient } from "./taxjar-client";
+import { TaxJarConfig } from "./taxjar-config";
+import { taxProviderUtils } from "../taxes/tax-provider-utils";
+
+type Payload = { order: OrderCreatedSubscriptionFragment; channel: ChannelConfig };
+type Target = CreateOrderArgs;
+type Response = CreateOrderResponse;
+
+export function sumPayloadLines(lines: LineItem[]): number {
+  return numbers.roundFloatToTwoDecimals(
+    lines.reduce(
+      (prev, line) =>
+        prev +
+        taxProviderUtils.resolveOptionalOrThrow(
+          line.unit_price,
+          new Error("Line unit_price is required to calculate order taxes")
+        ) *
+          taxProviderUtils.resolveOptionalOrThrow(
+            line.quantity,
+            new Error("Line quantity is required to calculate order taxes")
+          ),
+      0
+    )
+  );
+}
+
+export class TaxJarOrderCreatedPayloadTransformer {
+  private mapLines(lines: OrderCreatedSubscriptionFragment["lines"]): LineItem[] {
+    return lines.map((line) => ({
+      quantity: line.quantity,
+      unit_price: line.unitPrice.net.amount,
+      product_identifier: line.productSku ?? "",
+      // todo: add from tax code matcher
+      product_tax_code: "",
+      sales_tax: line.totalPrice.tax.amount,
+      description: line.productName,
+    }));
+  }
+
+  transform({ order, channel }: Payload): Target {
+    const lineItems = this.mapLines(order.lines);
+    const lineSum = sumPayloadLines(lineItems);
+    const shippingAmount = order.shippingPrice.gross.amount;
+    /**
+     * "The TaxJar API performs arbitrary-precision decimal arithmetic for accurately calculating sales tax."
+     * but we want to round to 2 decimals for consistency
+     */
```

Comment:
Are you sure we can do that? Not all currencies use 2 decimal places: https://docs.adyen.com/development-resources/currency-codes

Reply:
good catch, will investigate